ui/components/app/assets/token-list/token-list.tsx

```diff
@@ -1,56 +1,105 @@
-import useShouldShowFiat from '../hooks/useShouldShowFiat';
+import { useNativeTokenBalance, useNetworkFilter } from '../hooks';
+import { TokenWithFiatAmount } from '../types';
+import { getMultichainIsEvm } from '../../../../selectors/multichain';
+import { getConversionRate } from '../../../../ducks/metamask/metamask';
+import { filterAssets } from '../util/filter';
+import { sortAssets } from '../util/sort';
 
 type TokenListProps = {
   onTokenClick: (chainId: string, address: string) => void;
-  nativeToken?: ReactNode;
 };
 
-export default function TokenList({
-  onTokenClick,
-  nativeToken,
-}: TokenListProps) {
-  const { privacyMode } = useSelector(getPreferences);
+export default function TokenList({ onTokenClick }: TokenListProps) {
+  const isEvm = useSelector(getMultichainIsEvm);
   const chainIdsToPoll = useSelector(getChainIdsToPoll);
+  const newTokensImported = useSelector(getNewTokensImported);
+  const evmBalances = useSelector(getTokenBalancesEvm); // TODO: This is where we need to select non evm-assets from state, when isEvm is false
+  const currentNetwork = useSelector(getCurrentNetwork);
+  const { tokenSortConfig, privacyMode } = useSelector(getPreferences);
+  const selectedAccount = useSelector(getSelectedAccount);
+  const conversionRate = useSelector(getConversionRate);
+  const contractExchangeRates = useSelector(
+    getTokenExchangeRates,
+    shallowEqual,
+  );
 
   // EVM specific tokenBalance polling, updates state via polling loop per chainId
   pollAndUpdateEvmBalances({
     chainIds: chainIdsToPoll as Hex[],
   });
 
-  const sortedFilteredTokens = useSortedFilteredTokens();
-  const shouldShowFiat = useShouldShowFiat();
+  const nonEvmNativeToken = useNativeTokenBalance();
+
+  // network filter to determine which tokens to show in list
+  // on EVM we want to filter based on network filter controls, on non-evm we only want tokens from that chain identifier
+  const { networkFilter } = useNetworkFilter();
+
+  const sortedFilteredTokens = useMemo(() => {
+    const balances = isEvm ? evmBalances : [nonEvmNativeToken];
+    const filteredAssets = filterAssets(balances, [
+      {
+        key: 'chainId',
+        opts: isEvm ? networkFilter : { [nonEvmNativeToken.chainId]: true },
+        filterCallback: 'inclusive',
+      },
+    ]);
+
+    const { nativeTokens, nonNativeTokens } = filteredAssets.reduce<{
+      nativeTokens: TokenWithFiatAmount[];
+      nonNativeTokens: TokenWithFiatAmount[];
+    }>(
+      (acc, token) => {
+        if (token.isNative) {
+          acc.nativeTokens.push(token);
+        } else {
+          acc.nonNativeTokens.push(token);
+        }
+        return acc;
+      },
+      { nativeTokens: [], nonNativeTokens: [] },
+    );
+
+    const assets = [...nativeTokens, ...nonNativeTokens];
+
+    // sort filtered tokens based on the tokenSortConfig in state
+    return sortAssets(assets, tokenSortConfig);
```

Comment:
This is probably refactored/original logic, but it is a little confusing.

It seems like we split the list into 2:
```ts
const { nativeTokens, nonNativeTokens } = filteredAssets.reduce(...)
```

And then we just combined them again:
```
const assets = [...nativeTokens, ...nonNativeTokens]
return sortAssets(assets, tokenSortConfig)
```

Reply:
https://github.com/MetaMask/metamask-extension/pull/30238/commits/a4938d8d0cea570c52d4c24d9105534a8289cbf3

Seems like the reduce/join was a hack to satisfy e2e that was masking the root cause of the issue for when `tokenFiatAmount` was null, rather than a number like it expects. That commit handles the case when the value is nullish